tests/integration/records/record-data-test.js

```diff
@@ -0,0 +1,228 @@
+    rollbackAttribute() {}
+    changedAttributes() {}
+    hasChangedAttributes() {}
+    setAttr() {}
+    getAttr(member) {
+      return this.attributes !== null ? this.attributes[member] : undefined;
+    }
+    hasAttr(key) {
+      return key in this.attributes;
+    }
+    setHasMany() {}
+    getHasMany() {}
+    addToHasMany() {}
+    removeFromHasMany() {}
+    setBelongsTo() {}
+    getBelongsTo() {}
+
+    // missing from RFC
+    _initRecordCreateOptions(options) {
+      return options !== undefined ? options : {};
+    }
+
+    getResourceIdentifier() {
+      return {
+        id: this.id,
+        type: this.type,
+        clientId: this.clientId,
+      };
+    }
+    unloadRecord() {
+      this.attributes = null;
+      this.relationships = null;
+    }
+    isNew() {
+      return this.id === null;
+    }
+  }
+
+  test(`store.unloadRecord on a record with default RecordData with relationship to a record with custom RecordData does not error`, async function(assert) {
+    const originalCreateRecordDataFor = store.createRecordDataFor;
+    store.createRecordDataFor = function provideCustomRecordData(
+      modelName,
+      id,
+      clientId,
+      storeWrapper
+    ) {
+      if (modelName === 'pet') {
+        return new CustomRecordData(modelName, id, clientId, storeWrapper);
+      } else {
+        return originalCreateRecordDataFor.call(this, modelName, id, clientId, storeWrapper);
+      }
+    };
+
+    let chris = store.push({
+      data: {
+        type: 'person',
+        id: '1',
+        attributes: { name: 'Chris' },
+        relationships: {
+          pets: {
+            data: [{ type: 'pet', id: '1' }, { type: 'pet', id: '2' }],
+          },
+        },
+      },
+      included: [
+        {
+          type: 'pet',
+          id: '1',
+          attributes: { name: 'Shen' },
+          relationships: {
+            owner: { data: { type: 'person', id: '1' } },
+          },
+        },
+        {
+          type: 'pet',
+          id: '2',
+          attributes: { name: 'Prince' },
+          relationships: {
+            owner: { data: { type: 'person', id: '1' } },
+          },
+        },
+      ],
+    });
+    let pets = chris.get('pets');
+    let shen = pets.objectAt(0);
+
+    assert.equal(shen.get('name'), 'Shen', 'We found Shen');
+    assert.ok(
+      recordDataForRecord(chris) instanceof RecordData,
+      'We used the default record-data for person'
+    );
+    assert.ok(
+      recordDataForRecord(shen) instanceof CustomRecordData,
+      'We used the custom record-data for pets'
+    );
+
+    try {
+      run(() => chris.unloadRecord());
+      assert.ok(true, 'expected `unloadRecord()` not to throw');
+    } catch (e) {
+      assert.ok(false, 'expected `unloadRecord()` not to throw');
+    }
+  });
+
+  test(`store.unloadRecord on a record with custom RecordData with relationship to a record with default RecordData does not error`, async function(assert) {
```

Comment:
Let's add a test for DS.a -> custom -> DS.b; `DS.a.unloadRecord()` and confirm that this leaves `DS.b` around (unlike the non-custom link case)

Reply:
discussed oob that this test would be overly complicated for uncertain value